File: app/src/main/java/edu/gdei/gdeiassistant/Service/UpgradeService.java
package edu.gdei.gdeiassistant.Service;

import android.app.Notification;
import android.app.Service;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;

import edu.gdei.gdeiassistant.Constant.NotificationIDConstant;
import edu.gdei.gdeiassistant.NetWork.CheckUpgradeNetWork;
import edu.gdei.gdeiassistant.Pojo.JsonResult.DataJsonResult;
import edu.gdei.gdeiassistant.Pojo.Upgrade.CheckUpgradeResult;
import edu.gdei.gdeiassistant.R;

public class UpgradeService extends Service {

    @Override
    public void onCreate() {
        super.onCreate();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        //构建通知消息
        Notification notification = null;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(this, "service");
            notification = notificationBuilder
                    .setSmallIcon(R.drawable.info_small_icon).setBadgeIconType(R.mipmap.ic_launcher).setContentTitle("检查软件更新").setContentText("正在检查应用更新...")
                    .setWhen(System.currentTimeMillis()).setAutoCancel(true)
                    .build();
        } else {
            Notification.Builder builder = new Notification.Builder(this);
            builder.setSmallIcon(R.mipmap.ic_launcher).setContentTitle("检查软件更新").setContentText("正在检查应用更新...")
                    .setWhen(System.currentTimeMillis()).setAutoCancel(true);
            notification = builder.build();
        }
        //显示新服务的用户可见通知
        startForeground(NotificationIDConstant.UPGRADE_NOTIFICATION_ID, notification);
        //检查应用更新
        CheckUpgrade();
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        stopForeground(true);
    }

    /**
     * 检查更新，若有可用更新，则发送广播通知监听的Activity
     */
    private void CheckUpgrade() {
        new Thread() {
            @Override
            public void run() {
                super.run();
                try {
                    DataJsonResult<CheckUpgradeResult> result = new CheckUpgradeNetWork().GetUpgradeInformation(getApplicationContext());
                    if (result.isSuccess()) {
                        int currentVersionCode = getApplicationContext().getPackageManager().getPackageInfo(getApplicationContext().getPackageName(), 0).versionCode;
                        if (result.getData().getVersionCode() > currentVersionCode) {
                            //有新版本,提示更新
                            Intent intent = new Intent("edu.gdei.gdeiassistant.CHECK_UPGRADE");
                            if (result.getData().getVersionCodeName() != null) {
                                intent.putExtra("VersionCodeName", result.getData().getVersionCodeName());
                            }
                            if (result.getData().getDownloadURL() != null) {
                                intent.putExtra("DownloadURL", result.getData().getDownloadURL());
                            }
                            if (result.getData().getVersionInfo() != null) {
                                intent.putExtra("VersionInfo", result.getData().getVersionInfo());
                            }
                            if (result.getData().getFileSize() != null) {
                                intent.putExtra("FileSize", result.getData().getFileSize());
                            }
                            //发送广播通知更新
                            sendBroadcast(intent);
                        }
                    }
                    //检查更新完成后关闭服务
                    stopSelf();
                } catch (Exception ignored) {

                }
            }
        }.start();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/main/java/cn/gdeiassistant/service/UpgradeService.kt
package cn.gdeiassistant.service

import android.content.Intent
import android.os.IBinder
import androidx.core.app.NotificationCompat
import androidx.core.content.pm.PackageInfoCompat
import androidx.core.app.ServiceCompat
import androidx.lifecycle.LifecycleService
import androidx.lifecycle.lifecycleScope
import cn.gdeiassistant.constant.NotificationIDConstant
import cn.gdeiassistant.network.api.UpgradeApi
import cn.gdeiassistant.R
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

@AndroidEntryPoint
class UpgradeService : LifecycleService() {

    @Inject lateinit var upgradeApi: UpgradeApi

    override fun onCreate() { super.onCreate() }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val notification = NotificationCompat.Builder(this, "service")
            .setSmallIcon(R.drawable.ic_notification_info)
            .setBadgeIconType(NotificationCompat.BADGE_ICON_SMALL)
            .setContentTitle(getString(R.string.check_upgrade))
            .setContentText(getString(R.string.checking_update))
            .setWhen(System.currentTimeMillis())
            .setAutoCancel(true)
            .build()
        startForeground(NotificationIDConstant.UPGRADE_NOTIFICATION_ID, notification)
        checkUpgrade()
        return super.onStartCommand(intent, flags, startId)
    }

    override fun onDestroy() {
        super.onDestroy()
        ServiceCompat.stopForeground(this, ServiceCompat.STOP_FOREGROUND_REMOVE)
    }

    override fun onBind(intent: Intent): IBinder? = super.onBind(intent)

    private fun checkUpgrade() {
        lifecycleScope.launch {
            try {
                val result = withContext(Dispatchers.IO) { runCatching { upgradeApi.getUpgradeInfo() }.getOrNull() }
                if (result?.success == true && result.data != null) {
                    val currentVersionCode = PackageInfoCompat.getLongVersionCode(
                        packageManager.getPackageInfo(packageName, 0)
                    ).toInt()
                    val newCode = result.data!!.versionCode ?: 0
                    if (newCode > currentVersionCode) {
                        sendBroadcast(Intent("cn.gdeiassistant.CHECK_UPGRADE").apply {
                            result.data!!.versionCodeName?.let { putExtra("VersionCodeName", it) }
                            result.data!!.downloadURL?.let { putExtra("DownloadURL", it) }
                            result.data!!.versionInfo?.let { putExtra("VersionInfo", it) }
                            result.data!!.fileSize?.let { putExtra("FileSize", it) }
                        })
                    }
                }
            } catch (_: Exception) { }
            stopSelf()
        }
    }
}
